src/main/java/io/debezium/connector/spanner/db/metadata/SchemaRegistry.java

```diff
@@ -66,15 +67,28 @@ public synchronized TableSchema getWatchedTable(TableId tableId) {
 
     public synchronized void checkSchema(TableId tableId, Timestamp timestamp, List<Column> rowType) {
         if (!validate(tableId, rowType)) {
-            if (this.updateSchema(timestamp)) {
-                schemaResetTrigger.run();
-                return;
+            try {
+                if (this.updateSchema(timestamp)) {
+                    schemaResetTrigger.run();
+                    updateSchemaFromStaleTimestamp(tableId, timestamp, rowType);
+                    return;
+                }
+            }
+            catch (Exception e) {
+                // Only catch the exception due to read out of retention period
+                if (e.toString().contains("maximum timestamp staleness")) {
```

Comment:
Let's also catch the error code. I believe it is FAILED_PRECONDITION: https://cloud.google.com/spanner/docs/timestamp-bounds#maximum_timestamp_staleness

Also let's try for a bit longer of an error message

"